### ISSUE TITLE
[no-unsafe-subscript] doesn't catch some cases due to definite assignment analysis

### ISSUE THREAD

**Issue by danielnixon** [danielnixon/eslint-plugin-total-functions#68]
Observe this annoying case:

```ts
const arr = [0];
const foo: number | undefined = arr[0];
foo.toString(); // compiles; explodes at runtime
```

This is caused by [definite assignment analysis](https://www.reddit.com/r/typescript/comments/iffdlr/typescript_ignores_type_annotation_for_string/). Even though foo has an explicit type annotation of `number | undefined`, TypeScript determines that foo _cannot_ be undefined because it is "definitely assigned". `toString()` then explodes at runtime.

Our no-unsafe-subscript rule currently _doesn't_ flag `const foo: number | undefined = arr[0]` because the [contextual type](https://www.typescriptlang.org/docs/handbook/type-inference.html#contextual-typing) of the expression `arr[0]` is `number | undefined` (the type of the value it's being assigned to).

This demonstrates that we _can't_ assume the array access is safe even when the contextual type contains `undefined` -- at least for assignment expressions, which as far as I can tell are the only place that definite assignment analysis applies.

We will have to update no-unsafe-subscript to either ignore the contextual type in all cases (which would cause lots of pain and what amount to false positives) or only in assignment expressions (if possible).

🤦

**Comment by danielnixon** [comment]
Released as v4.0.0
